### ISSUE TITLE
训练时，验证集的效果非常差是怎么回事

### ISSUE THREAD

**Issue by sxxtaotao** [leeyeehoo/CSRNet-pytorch#48]
当我训练的时候，验证集的效果非常差，一直在300-400左右浮动，请问这是什么原因，我也没有改别的，只是把python3中不兼容python2的给改掉了，改动最大的是model.py中
 if not load_weights:
            mod = models.vgg16(pretrained = True)
            self._initialize_weights()           
            #读取参数
            pretrained_dict=mod.state_dict()
            self.frontend_dict=self.frontend.state_dict()
            #将pretrained_dict里不属于frontend_dict的键剔除掉
            pretrained_dict={k:v for k, v in pretrained_dict.items() if k in self.frontend_dict}
            #更新现有的frontend_dict
            self.frontend_dict.update(pretrained_dict)
            #加载我们真正需要的state_dict
            self.frontend.load_state_dict(self.frontend_dict)

因为python3中不支持这样的索引，不知道改的对不对，验证集效果一直很差

**Comment by neverland0621** [comment]
我猜你验证集的代码不对，记得最后一个cell中有一行注释，取消掉就好了

**Comment by Garyzdz** [comment]
我的预测值为nan

**Comment by zhengmiao1** [comment]
> 我猜你验证集的代码不对，记得最后一个cell中有一行注释，取消掉就好了

你说的挺对的，我开始也是验证集精度很差，验证集part_B 的MAE是1900多，取消掉最后一个cell中的注释img = transform(Image.open(img_paths[i]).convert('RGB')).cuda()后，我的验证集精度为8.998，感谢

**Comment by Cli98** [comment]
> > 我猜你验证集的代码不对，记得最后一个cell中有一行注释，取消掉就好了
> 
> 你说的挺对的，我开始也是验证集精度很差，验证集part_B 的MAE是1900多，取消掉最后一个cell中的注释img = transform(Image.open(img_paths[i]).convert('RGB')).cuda()后，我的验证集精度为8.998，感谢

If you do this, then you negate all normalization process before this line.

**Comment by Cli98** [comment]
> 当我训练的时候，验证集的效果非常差，一直在300-400左右浮动，请问这是什么原因，我也没有改别的，只是把python3中不兼容python2的给改掉了，改动最大的是model.py中
> if not load_weights:
> mod = models.vgg16(pretrained = True)
> self._initialize_weights()
> #读取参数
> pretrained_dict=mod.state_dict()
> self.frontend_dict=self.frontend.state_dict()
> #将pretrained_dict里不属于frontend_dict的键剔除掉
> pretrained_dict={k:v for k, v in pretrained_dict.items() if k in self.frontend_dict}
> #更新现有的frontend_dict
> self.frontend_dict.update(pretrained_dict)
> #加载我们真正需要的state_dict
> self.frontend.load_state_dict(self.frontend_dict)
> 
> 因为python3中不支持这样的索引，不知道改的对不对，验证集效果一直很差

你改错了
<img width="542" alt="image" src="https://user-images.githubusercontent.com/19630616/73401752-bddd4780-42b9-11ea-8367-c18b8d848974.png">

**Comment by Cli98** [comment]
> > 当我训练的时候，验证集的效果非常差，一直在300-400左右浮动，请问这是什么原因，我也没有改别的，只是把python3中不兼容python2的给改掉了，改动最大的是model.py中
> > if not load_weights:
> > mod = models.vgg16(pretrained = True)
> > self._initialize_weights()
> > #读取参数
> > pretrained_dict=mod.state_dict()
> > self.frontend_dict=self.frontend.state_dict()
> > #将pretrained_dict里不属于frontend_dict的键剔除掉
> > pretrained_dict={k:v for k, v in pretrained_dict.items() if k in self.frontend_dict}
> > #更新现有的frontend_dict
> > self.frontend_dict.update(pretrained_dict)
> > #加载我们真正需要的state_dict
> > self.frontend.load_state_dict(self.frontend_dict)
> > 因为python3中不支持这样的索引，不知道改的对不对，验证集效果一直很差
> 
> 你改错了
> <img alt="image" width="542" src="https://user-images.githubusercontent.com/19630616/73401752-bddd4780-42b9-11ea-8367-c18b8d848974.png">

Python 3 应该这么改
        if not load_weights:
            mod = models.vgg16(pretrained = True)
            self._initialize_weights()
            pre_train_item = list(mod.state_dict().items())
            # reload as much as pre-trained weight as possible
            for i in range(len(self.frontend.state_dict().items())):
                pre_train_weight_name, pre_train_weight_val = \
                    pre_train_item[i][0].replace("features.",""), pre_train_item[i][1].data[:]
                self.frontend.state_dict()[pre_train_weight_name] = pre_train_weight_val
